File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/playlist_screens/playlist/adapter/PlayListItemAdapter.java
package com.github.anrimian.musicplayer.ui.playlist_screens.playlist.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.github.anrimian.musicplayer.domain.models.playlist.PlayListItem;
import com.github.anrimian.musicplayer.domain.models.utils.PlayListItemHelper;
import com.github.anrimian.musicplayer.domain.utils.functions.BiCallback;
import com.github.anrimian.musicplayer.ui.utils.OnItemClickListener;
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.diff_utils.SimpleDiffItemCallback;
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.diff_utils.adapter.DiffListAdapter;

/**
 * Created on 31.10.2017.
 */

public class PlayListItemAdapter extends DiffListAdapter<PlayListItem, PlayListItemViewHolder> {

    private final boolean coversEnabled;

    private final BiCallback<PlayListItem, Integer> onCompositionClickListener;
    private final OnItemClickListener<Integer> onIconClickListener;

    public PlayListItemAdapter(RecyclerView recyclerView,
                               boolean coversEnabled,
                               BiCallback<PlayListItem, Integer> onCompositionClickListener,
                               OnItemClickListener<Integer> onIconClickListener) {
        super(recyclerView, new SimpleDiffItemCallback<>(
                PlayListItemHelper::areSourcesTheSame,
                PlayListItemHelper::getChangePayload)
        );
        this.coversEnabled = coversEnabled;
        this.onCompositionClickListener = onCompositionClickListener;
        this.onIconClickListener = onIconClickListener;
    }

    @NonNull
    @Override
    public PlayListItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new PlayListItemViewHolder(LayoutInflater.from(parent.getContext()),
                parent,
                onCompositionClickListener,
                onIconClickListener);
    }

    @Override
    public void onBindViewHolder(@NonNull PlayListItemViewHolder holder, int position) {
        PlayListItem composition = getItem(position);
        holder.bind(composition, coversEnabled);
    }

    @Override
    public void onViewRecycled(@NonNull PlayListItemViewHolder holder) {
        super.onViewRecycled(holder);
        holder.release();
    }
}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/playlist_screens/playlist/adapter/PlayListItemAdapter.kt
package com.github.anrimian.musicplayer.ui.playlist_screens.playlist.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.RecyclerView
import com.github.anrimian.filesync.models.state.file.FileSyncState
import com.github.anrimian.musicplayer.domain.models.composition.CurrentComposition
import com.github.anrimian.musicplayer.domain.models.playlist.PlayListItem
import com.github.anrimian.musicplayer.domain.models.utils.PlayListItemHelper
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.diff_utils.SimpleDiffItemCallback
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.mvp.MvpDiffAdapter

/**
 * Created on 31.10.2017.
 */
class PlayListItemAdapter(
    lifecycleOwner: LifecycleOwner,
    recyclerView: RecyclerView,
    private val coversEnabled: Boolean,
    private val menuClickListener: (View, Int, PlayListItem) -> Unit,
    private val onItemClickListener: (PlayListItem, Int) -> Unit
) : MvpDiffAdapter<PlayListItem, PlayListItemViewHolder>(
    lifecycleOwner,
    recyclerView,
    SimpleDiffItemCallback(PlayListItemHelper::areSourcesTheSame, PlayListItemHelper::getChangePayload)
) {

    private var currentComposition: CurrentComposition? = null
    private var syncStates = emptyMap<Long, FileSyncState>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PlayListItemViewHolder {
        return PlayListItemViewHolder(
            LayoutInflater.from(parent.context),
            parent,
            menuClickListener,
            onItemClickListener
        )
    }

    override fun onBindViewHolder(holder: PlayListItemViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)
        val composition = getItem(position)
        holder.bind(composition, coversEnabled)
        holder.showCurrentComposition(currentComposition, false)
        holder.setFileSyncStates(syncStates)
    }

    fun showCurrentComposition(currentComposition: CurrentComposition?) {
        this.currentComposition = currentComposition
        forEachHolder { holder ->
            holder.showCurrentComposition(currentComposition, true)
        }
    }

    fun showFileSyncStates(states: Map<Long, FileSyncState>) {
        this.syncStates = states
        forEachHolder { holder ->
            holder.setFileSyncStates(syncStates)
        }
    }
}